### ISSUE TITLE
Debug rosetta-cli balance reconciliation.

### ISSUE THREAD

**Issue by Sventimir** [MinaProtocol/mina#12276]
When run against a very simple sandbox blockchain, including only a single transaction, `rosetta-cli` [tool](https://github.com/coinbase/rosetta-cli) reports reconciliation errors. In short this means that when `rosetta-cli` tries to track balances of various accouts, the expected balance it computes differs from what our Rosetta implementation reports. I have investigated the issue thoroughly and it appears that Rosetta is correct in its reports and its `rosetta-cli` which gets the expected balances wrong. In particular it appears to be taking at least some of the transactions into account twice. This looks like a bug in `rosetta-cli`, which needs to be investigated.

I tried asking a question on [community forum](https://community.rosetta-api.org/t/how-does-rosetta-cli-actually-track-balances/831), but got no response so far. So I'm debugging this myself.

**Comment by Sventimir** [comment]
I located the place in `rosetta-cli`'s code, where the in-memory balance is being updated. Against my previous suspicion, I discovered that the balance is only updated once, but it is being updated based on the balance *after* the transaction and adds (or subtracts) another transaction's amount to the balance. This explains why transactions are being taken into account twice. However, the root cause of the error is still not known. There are several possibilities:
* another thread erroneously updates the balance as well in a place I didn't locate yet;
* `rosetta-cli` interprets balance reported by Rosetta to be before the transaction, while Rosetta actually reports the balance after the transaction;
* there's an issue with balance initialisation in `rosetta-cli`.

In other words, it's possible that our Rosetta implementation reports initial balances in an unexpected way, but still a bug in `rosetta-cli` itself seems more plausible, because all responses from our Rosetta implementation seem to be in accordance with the specification.

**Comment by Sventimir** [comment]
Having gathered enough understanding of how `rosetta-cli` tracks balances, I realised that the bug was in our implementation all the time. More precisely, when asked for an account's balance at a particular block, before any transactions were made to that account, Rosetta returns the most recent balance instead. Because `rosetta-cli` relies on Rosetta to gather initial balances of accounts and these are precisely those balances that are misreported, we got the reconciliation error.
